File: src/Game/ReceiveReplication.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Components = require(ReplicatedStorage.Shared.Components)
local RemoteEvent = ReplicatedStorage:WaitForChild("MatterRemote")

local Codecs = require(ReplicatedStorage.Shared.Codecs)

local function SetupReplication(world, state)
    local function debugPrint(...)
        if state.debugEnabled then
            print("Replication>", ...)
        end
    end

    local entityIdMap = {}

    RemoteEvent.OnClientEvent:Connect(function(entities)
        for serverEntityId, componentMap in entities do
            local clientEntityId = entityIdMap[serverEntityId]

            if clientEntityId and next(componentMap) == nil then
                world:despawn(clientEntityId)
                entityIdMap[serverEntityId] = nil
                debugPrint(string.format("Despawn %ds%d", clientEntityId, serverEntityId))
                continue
            end

            local componentsToInsert = {}
            local componentsToRemove = {}

            local insertNames = {}
            local removeNames = {}

            for name, container in componentMap do
                -- If we have encoded any data then first we want to decode it
                -- before adding it to the component.
                if Codecs[name] then
                    container.data = Codecs[name].decode(container.data)
                end

                if container.data then
                    table.insert(componentsToInsert, Components[name](container.data))
                    table.insert(insertNames, name)
                else
                    table.insert(componentsToRemove, Components[name])
                    table.insert(removeNames, name)
                end
            end

            if clientEntityId == nil then
                clientEntityId = world:spawn(unpack(componentsToInsert))

                entityIdMap[serverEntityId] = clientEntityId

                debugPrint(
                    string.format("Spawn %ds%d with %s", clientEntityId, serverEntityId, table.concat(insertNames, ","))
                )
            else
                if #componentsToInsert > 0 then
                    world:insert(clientEntityId, unpack(componentsToInsert))
                end

                if #componentsToRemove > 0 then
                    world:remove(clientEntityId, unpack(componentsToRemove))
                end

                debugPrint(
                    string.format(
                        "Modify %ds%d adding %s, removing %s",
                        clientEntityId,
                        serverEntityId,
                        if #insertNames > 0 then table.concat(insertNames, ", ") else "nothing",
                        if #removeNames > 0 then table.concat(removeNames, ", ") else "nothing"
                    )
                )
            end
        end
    end)
end

return SetupReplication

File: src/client/receive-replication.ts
import { AnyComponent, AnyEntity, World } from "@rbxts/matter";
import { ComponentCtor } from "@rbxts/matter/lib/component";
import { codecs } from "shared/codecs";
import * as Components from "shared/components";
import Remotes from "shared/remotes";
import { UnionComponentsMap } from "shared/serde";

const remoteEvent = Remotes.Client.Get("Replication");

function receiveReplication(world: World): void {
	const entityIdMap = new Map<string, AnyEntity>();

	remoteEvent.Connect((entities) => {
		for (const [serverEntityId, componentMap] of entities) {
			let clientEntityId = entityIdMap.get(serverEntityId);

			// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
			if (clientEntityId !== undefined && next(componentMap)[0] === undefined) {
				world.despawn(clientEntityId);
				entityIdMap.delete(serverEntityId);
				continue;
			}

			const componentsToInsert = new Array<AnyComponent>();
			const componentsToRemove = new Array<ComponentCtor>();

			for (const [name, container] of componentMap) {
				// If we have encoded any data then first we want to decode it
				// before adding it to the component.
				const codec = codecs.get(name);
				if (codec !== undefined && container.data !== undefined) {
					container.data = codec.decode(container.data);
				}

				if (container.data !== undefined) {
					componentsToInsert.push(Components[name](container.data as UnionComponentsMap));
				} else {
					componentsToRemove.push(Components[name]);
				}
			}

			if (clientEntityId === undefined) {
				clientEntityId = world.spawn(...componentsToInsert);
				entityIdMap.set(serverEntityId, clientEntityId);
			} else {
				if (componentsToInsert.size() > 0) {
					world.insert(clientEntityId, ...componentsToInsert);
				}

				if (componentsToRemove.size() > 0) {
					world.remove(clientEntityId, ...componentsToRemove);
				}
			}
		}
	});
}

export default receiveReplication;
